lib/ansible/galaxy/api.py

```diff
@@ -696,17 +697,26 @@ def wait_import_task(self, task_id, timeout=0):
         start = time.time()
         wait = 2
 
+        token_expired = False
         while timeout == 0 or (time.time() - start) < timeout:
             try:
                 data = self._call_galaxy(full_url, method='GET', auth_required=True,
                                          error_context_msg='Error when getting import task results at %s' % full_url)
             except GalaxyError as e:
-                if e.http_code != 404:
+                if e.http_code == 401 and self.token and isinstance(self.token, KeycloakToken):
+                    if token_expired:
+                        raise
+                    display.vvv("Authentication failed, refreshing token to try again.")
+                    self.token._token = None
+                    token_expired = True
```

Comment:
I think it'd be more organic to implement a `KeycloakToken.expired` property. The `KeycloakToken.get()` method would need to get updated to store the expiration time.
And then, there could be an `KeycloakToken.expire()` method for resetting the token.

I'd also recommend adding a skew offset to the `KeycloakToken.expired` property. So it'd "expire" in on the client side slightly earlier than it does on the server.
This could later be used in the API call wrapper to pre-refresh the token if it's about to expire. A refresh call could be added before this line https://github.com/ansible/ansible/blob/f02b8fb/lib/ansible/galaxy/api.py#L428 to update the token if it's turning stale.

This proactive approach would make the operation smoother than the reactive one, most of the time.

Reply:
Good idea. Rather than an `expire` method to update the token, `KeycloakToken.get()` handles the refresh for expired tokens. `KeycloakToken.headers()` uses `KeycloakToken.get()` as well, so this should all be implemented now.